gcalbot/db.sql

```diff
@@ -1,20 +1,63 @@
-DROP TABLE IF EXISTS `accounts`;
+DROP TABLE IF EXISTS `account`;
 
-CREATE TABLE `accounts` (
-  `username` varchar(128) NOT NULL,
-  `nickname` varchar(128) NOT NULL,
-  PRIMARY KEY (`username`, `nickname`)
+CREATE TABLE `account` (
+    `account_id` varchar(128) NOT NULL,         -- unique id per connected google account (currently kbuser:nickname)
+    `keybase_username` varchar(128) NOT NULL,   -- kb username
+    `account_nickname` varchar(128) NOT NULL,   -- nickname of google account for kb user
+    PRIMARY KEY (`account_id`),
+    UNIQUE KEY (`keybase_username`, `account_nickname`)
 ) ENGINE=InnoDB DEFAULT CHARSET=utf8;
 
 DROP TABLE IF EXISTS `oauth`;
 
 CREATE TABLE `oauth` (
-  `identifier` varchar(128) UNIQUE NOT NULL,
-  `ctime` datetime NOT NULL,
-  `mtime` datetime NOT NULL,
-  `access_token` varchar(256) NOT NULL,
-  `token_type` varchar(64) NOT NULL,
-  `refresh_token` varchar(256) NOT NULL,
-  `expiry` datetime NOT NULL,
-  PRIMARY KEY (`identifier`)
+    `identifier` varchar(128) UNIQUE NOT NULL,
+    `ctime` datetime NOT NULL,
+    `mtime` datetime NOT NULL,
+    `access_token` varchar(256) NOT NULL,
+    `token_type` varchar(64) NOT NULL,
+    `refresh_token` varchar(256) NOT NULL,
+    `expiry` datetime NOT NULL,
+    PRIMARY KEY (`identifier`)
+    -- oauth row is created before account row, so currently can't have foreign key reference to account
```

Comment:
I do think it would simplify the code and db, but it makes the auth process a bit more complicated. Right now the current process is as follows:
1. The user runs `!gcal accounts connect work` or a command such as `!gcal list-calendars work`
2. The bot sends a oauth link to the user
3. When the user authenticates, the oauth token is inserted into the db
4. The oauth callback is run, and the account is inserted into the db (after the oauth token)

Currently I have two solutions in mind:
1. Insert the account into the db after step 1, before the oauth link is sent. This would add some complexity, because if accounts now have a state of being unauthenticated (as in the user hasn't clicked the oauth link yet)
2. Do some custom oauth stuff. The abstractions are ok, but it's sometimes tough to blend in the concept of "multiple accounts" with it.

Reply:
Couldn't 3. and 4. be swapped in the abstraction?

Reply:
The abstraction doesn't have any concept of the account table that I have, and the callback should be run after the oauth token is inserted

Reply:
Other option could be the callback takes in a transaction so the callback can operate on the database in the same tx as the oauth

Reply:
@joshblum

Reply:
yes, the insert account would have to precde the oauth insert, just syaing they could be contained within the tx together